common/astyanax/src/main/java/com/bazaarvoice/emodb/common/cassandra/cqldriver/AdaptiveResultSet.java

```diff
@@ -0,0 +1,303 @@
+        if (!isAdaptiveException(reason) || --_remainingAdaptations == 0) {
+            return false;
+        }
+
+        ExecutionInfo executionInfo = _delegate.getExecutionInfo();
+        Statement statement = executionInfo.getStatement();
+        PagingState pagingState = executionInfo.getPagingState();
+        int fetchSize = statement.getFetchSize();
+
+        while (fetchSize > MIN_FETCH_SIZE) {
+            fetchSize = Math.max(fetchSize / 2, MIN_FETCH_SIZE);
+            _log.debug("Retrying query at next page with fetch size {} due to {}", fetchSize, reason.getMessage());
+            statement.setFetchSize(fetchSize);
+            statement.setPagingState(pagingState);
+            try {
+                _delegate = _session.execute(statement);
+                return true;
+            } catch (Throwable t) {
+                // Exit the adaptation loop if the exception isn't one where adapting further may help
+                if (!isAdaptiveException(t) || --_remainingAdaptations == 0) {
+                    return false;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    public Iterator<Row> iterator() {
+        return new AbstractIterator<Row>() {
+            @Override
+            protected Row computeNext() {
+                Row next = one();
+                if (next != null) {
+                    return next;
+                }
+                return endOfData();
+            }
+        };
+    }
+
+    @Override
+    public List<Row> all() {
+        return StreamSupport.stream(spliterator(), false).collect(Collectors.toList());
+    }
+
+    @Override
+    public ListenableFuture<ResultSet> fetchMoreResults() {
+        final ResultSet delegate = _delegate;
+
+        // If we've already tried to pre-fetch for this delegate and ran into frame size issues then don't try again.
+        if (_delegateWithPrefetchFailure == delegate) {
+            return Futures.immediateFuture(this);
+        }
+
+        // Change the returned future to contain this instance instead of the delegate
+        ListenableFuture<ResultSet> future = Futures.transform(delegate.fetchMoreResults(), new Function<ResultSet, ResultSet>() {
+            @Override
+            public ResultSet apply(ResultSet ignore) {
+                return AdaptiveResultSet.this;
+            }
+        });
+
+        Futures.addCallback(future, new MoreFutures.FailureCallback<ResultSet>() {
+            @Override
+            public void onFailure(Throwable t) {
+                // The async pre-fetch has failed.  Check if the root cause is adaptive.
+                if (isAdaptiveException(t)) {
+                    // Future:  Optimize to pre-fetch the next delegate immediately.  For now simply record that
+                    // we shouldn't try to pre-fetch again for this delegate.  The frame size will be adjusted
+                    // synchronously after all available rows have been consumed.
```

Comment:
This is a bit of a cop-out.  It wouldn't be too hard to maintain two delegates: the "old" one for serving up fetched rows and a "new" one for serving the remaining unfetched rows.  The real trick is coordinating potentially synchronous (via `one()`) and asynchronous (via `fetchMoreResults()`) operations which both would race to create a new delegate instance.  This, too, isn't too hard, but these changes would take this already complex pull request and add even more unnecessary complexity for a very slight optimization.  The solution here is to just defer to the synchronous call if the asynchronous one fails and save the optimization for another day.